compiler/src/iree/compiler/Utils/InterfaceUtils.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023 The IREE Authors
```

Comment:
"Interface" still doesn't make sense and is confusing in other ways now since it looks like MLIR Op/AttrInterface utils.

All of this logic doesn't relate to interfaces - it's changing data types and encodings. ElementPackingUtils or something, maybe, but I'm skeptical of this being in Utils/.

I'm unsure of why TypePropagationPass happens in Codegen/ - all of this logic should be limited to the stream dialect where it performs these operations - after encoding the tensors the program should be updated to be in a consistent state and there should be no need to then _also_ do some fixup later on in each codegen backend that reaches back up the stack into what _may_ have happened earlier in the stream dialect (the stream dialect is not a requirement for codegen, and codegen is not a requirement for the stream dialect). If the TypePropagationPass needs to run on executables then it can run after the encoding pass in the stream pipeline nested on `IREE::Stream::ExecutableOp`s.

Reply:
I got the interface term from vulkan shader interfaces and such: https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap15.html#interfaces. But yeah I see it can be potentially confusing with mlir interfaces.. So switched to use "storage", which is a more neural term and reflects the purpose better I guess.

Regarding  `TypePropagationPass`, I don't have full context on it; maybe @MaheshRavishankar can shed some lights. I think the intent was that different backends might have different views w.r.t. of whether a type is legal or not? Though right now the pass is not configurable and it's applied for all backends unconditionally.  I'm happy to clean up this part if we think it should be moved to stream instead.

Reply:
> I got the interface term from vulkan shader interfaces and such: [registry.khronos.org/vulkan/specs/1.3-extensions/html/chap15.html#interfaces](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap15.html#interfaces). But yeah I see it can be potentially confusing with mlir interfaces.. So switched to use "storage", which is a more neural term and reflects the purpose better I guess.
> 
> Regarding `TypePropagationPass`, I don't have full context on it; maybe @MaheshRavishankar can shed some lights. I think the intent was that different backends might have different views w.r.t. of whether a type is legal or not? Though right now the pass is not configurable and it's applied for all backends unconditionally. I'm happy to clean up this part if we think it should be moved to stream instead.

Might benefit from some high-bandwidth communication... but when this was added, the stream dialect would insert some `i8` to `i1` casts at the boundaries. This pass is propagating those within the dispatch to not end up with `i1` loads (so propagate the types across Linalg ops basically). We can move that out of Codegen. It is run before any tiling, etc. is run....

Reply:
Filed #13520 to track potential clarification/cleanup. I think since this PR is only making this pass more complex but not changing where/when it's run it's fine to land as-is.